File: Editorial/2_Add_Two_Numbers/Java/3_In_Place_Addition.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = l1 != null ? l1 : l2;
        ListNode current = head;
        ListNode previous = null;
        int carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            if (current == null) {
                current = new ListNode(0);
                previous.next = current;
            }

            int total = carry;
            if (l1 != null) {
                total += l1.val;
                l1 = l1.next;
            }

            if (l2 != null) {
                total += l2.val;
                l2 = l2.next;
            }

            current.val = total % 10;
            carry = total / 10;
            previous = current;

            if (current.next != null) {
                current = current.next;
            } else if (l1 != null) {
                current.next = l1;
                current = current.next;
            } else if (l2 != null) {
                current.next = l2;
                current = current.next;
            } else {
                current = null;
            }
        }

        return head;
    }
}

File: Editorial/2_Add_Two_Numbers/Go/3_In_Place_Addition.go
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    head := l1
    if head == nil {
        head = l2
    }

    current := head
    var previous *ListNode
    carry := 0

    for l1 != nil || l2 != nil || carry != 0 {
        if current == nil {
            current = &ListNode{}
            previous.Next = current
        }

        total := carry
        if l1 != nil {
            total += l1.Val
            l1 = l1.Next
        }

        if l2 != nil {
            total += l2.Val
            l2 = l2.Next
        }

        current.Val = total % 10
        carry = total / 10
        previous = current

        if current.Next != nil {
            current = current.Next
        } else if l1 != nil {
            current.Next = l1
            current = current.Next
        } else if l2 != nil {
            current.Next = l2
            current = current.Next
        } else {
            current = nil
        }
    }

    return head
}
